invenio_github/models.py

```diff
@@ -129,6 +285,58 @@ class Release(db.Model, Timestamp):
     )
     """Record identifier."""
 
-    repository = db.relationship(Repository, backref='releases')
+    status = db.Column(
+        ChoiceType(ReleaseStatus, impl=db.CHAR(1)),
+        nullable=False,
+    )
+    """Status of the release, e.g. 'processing', 'published', 'failed', etc."""
+
+    repository = db.relationship(
+        Repository,
+        backref=db.backref('releases', lazy='dynamic')
+    )
     record = db.relationship(RecordMetadata)
     event = db.relationship(Event)
+
+    @classmethod
+    def create(cls, event):
+        """Create a new Release model."""
+        release_id = event.payload['release']['id']
+
+        # Check if the release has already been received
+        existing_release = Release.query.filter_by(
+            release_id=release_id,
+        ).first()
+        if existing_release:
+            msg = ('Release [{tag}:{release_id}] has already been received '
+                   'with status {status}.')
+            raise Exception(msg.format(tag=existing_release.tag,
+                                       release_id=release_id,
+                                       status=existing_release.status.title))
+
+        # Check if the release corresponds to a repository in our database
+        repo_id = event.payload['repository']['id']
+        repository = Repository.get(user_id=event.user_id, github_id=repo_id)
+        if repository:
+            # Create the Release
+            with db.session.begin_nested():
+                release = cls(
+                    release_id=release_id,
+                    tag=event.payload['release']['tag_name'],
+                    repository=repository,
+                    event=event,
+                    status=ReleaseStatus.RECEIVED,
+                )
+                db.session.add(release)
+            return release
+        else:
+            msg = 'Repository [{repo_name}:{repo_id}] is not enabled.'
+            raise Exception(
+                msg.format(repo_name=event.payload['repository']['full_name'],
+                           repo_id=repo_id))
+
+    @property
+    def doi(self):
+        """Get DOI of Release from record metadata."""
+        if self.record:
+            return self.record.json.get('doi')
```

Comment:
- This should be configurable - shall we use JSON Path?
- IMHO we should move it from the model to API.